src/lightning/fabric/strategies/model_parallel.py

```diff
@@ -0,0 +1,330 @@
+        if isinstance(state, Module) or isinstance(state, Optimizer):
+            raise NotImplementedError(
+                "Loading a module or optimizer object from a checkpoint directly is not yet supported."
+            )
+        if strict is False:
+            raise NotImplementedError(f"Non-strict loading is not yet supported in {self.__class__.__name__}.")
+
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        path = Path(self.broadcast(path))
+        _distributed_checkpoint_load(state, path)  # type: ignore[arg-type]
+        return {}
+
+    def _setup_distributed(self) -> None:
+        reset_seed()
+        self._set_world_ranks()
+        self._process_group_backend = self._get_process_group_backend()
+        assert self.cluster_environment is not None
+        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+
+    def _setup_device_mesh(self) -> None:
+        if self._data_parallel_size == "auto":
+            self._data_parallel_size = self.num_nodes
+        if self._tensor_parallel_size == "auto":
+            self._tensor_parallel_size = self.num_processes
+        if self._data_parallel_size * self._tensor_parallel_size != self.world_size:
+            raise RuntimeError(
+                f"The sizes `data_parallel_size={self._data_parallel_size}` and"
+                f" `tensor_parallel_size={self._tensor_parallel_size}` multiplied should equal the world size"
+                f" ({self.world_size})."
+            )
+        self._device_mesh = init_device_mesh(
+            device_type=self.root_device.type,
+            mesh_shape=(self._data_parallel_size, self._tensor_parallel_size),
+            mesh_dim_names=("data_parallel", "tensor_parallel"),
+        )
+
+    def _get_process_group_backend(self) -> str:
+        return self._process_group_backend or _get_default_process_group_backend_for_device(self.root_device)
+
+    def _set_world_ranks(self) -> None:
+        if self.cluster_environment is not None:
+            self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+            self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        # `LightningEnvironment.set_global_rank` will do this too, but we cannot rely on that implementation detail
+        # additionally, for some implementations, the setter is a no-op, so it's safer to access the getter
+        rank_zero_only.rank = utils_rank_zero_only.rank = self.global_rank
+
+
+def _materialize_module(module: Module, device: torch.device) -> None:
+    # Reference: https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md#meta-device-initialization
+    # TODO: Introduce `Fabric.materialize(module)` to give user control when materialization should happen
+    # TODO: Make `torchmetrics.Metric` compatible with the `to_empty()` + `reset_parameters()` semantics
```

Comment:
Not incredibly high priority for Fabric, but it will be relevant for LightningModules where users typically put TorchMetrics as attributes, and I'll reuse this function in the Trainer. I'll open a PR for this shortly.